nixfmt[-rfc-style]: unstable -> 1.0.0

infinisil opened the pull request:
Updates Nixfmt to the stable release, see https://github.com/NixOS/nixfmt/pull/316 and https://github.com/NixOS/nixfmt/issues/272.

Cc @GaetanLepage ;)

Also:
- Updates the update script to use stable versions going forward
- Make pkgs.nixfmt the -rfc-style version and remove the warning
- Create a (delayed) warning for the -rfc-style version to encourage switching to pkgs.nixfmt in a couple releases

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

This work is funded by [Antithesis](https://antithesis.com/) and [Tweag](https://tweag.io) :sparkles:

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

infinisil committed:
nixfmt[-rfc-style]: unstable -> 1.0.0

Also:
- Updates the update script to use stable versions going forward
- Make pkgs.nixfmt the -rfc-style version and remove the warning
- Create a (delayed) warning for the -rfc-style version to encourage
  switching to pkgs.nixfmt in a couple releases
- Add a release note for the above

acid-bong commented:
yeah, i prefer yours 🌚

infinisil committed:
treewide: nixfmt-rfc-style -> nixfmt

Except:
- Instances in documentation, because people in older versions
  can't switch to nixfmt yet due to it having pointed to nixfmt-classic
  before
- In code that runs based on a CI Nixpkgs version, which is also a bit
  older still
- In update script shebangs, because many of them don't pin Nixpkgs, and run
  with whatever is in NIX_PATH (and it's not easy to fix this, see
  https://github.com/NixOS/nixpkgs/issues/425551)

infinisil commented:
Yeah, once we update the pinned CI tooling (once the result of this PR is built by Hydra), we'll need to reformat Nixpkgs again and backport that.

wolfgangwalther reviewed (approved):
Congratz on 1.0.0!

GaetanLepage reviewed (approved):
Well done!

MattSturgeon commented on maintainers/scripts/haskell/regenerate-hackage-packages.sh:
```diff
@@ -1,5 +1,5 @@
 #! /usr/bin/env nix-shell
-#! nix-shell -i bash -p coreutils haskellPackages.cabal2nix-unstable git nixfmt-rfc-style -I nixpkgs=.
+#! nix-shell -i bash -p coreutils haskellPackages.cabal2nix-unstable git nixfmt -I nixpkgs=.
```
This treewide highlights that a lot of update scripts are using nixfmt from `.` instead of the nixpkgs revision pinned in `ci/`.

@emilazy is right that these update scripts will format using a different version of nixfmt to that used by CI. This isn't an issue with this PR, but this PR does highlight the issue.

Should we open a tracking issue for migrating update scripts to use correctly pinned dependencies?

MattSturgeon replied:
Yeah, I guess in many cases this would mean having _multiple_ `nix-shell` shebangs. In my testing that seems to work fine.

In this instance, we'd need one line installing packages from `nixpkgs=.`, and another line entering the `ci` shell or installing nixfmt from `./ci`.

Currently `nix-shell ci -A shell` provides access to `treefmt` and `nixfmt`, so that may be a bit overkill if most of theses update scripts don't need our configured treefmt and all its formatters...

---

Since (AFAIK) `-p` can only work by using `nixpkgs` from the lookup path and `-A` only works by treating the attribute as a shell, I guess we'd need add a shell to `ci/default.nix` that makes nixfmt available? Or add an attribute to the main pkgs set that provides access to the pinned nixpkgs revision?

That way we could do something like:

```sh
#! /usr/bin/env nix-shell
#! nix-shell -i bash -p coreutils haskellPackages.cabal2nix-unstable git -I nixpkgs=.
#! nix-shell -i bash -p pinned.nixfmt -I nixpkgs=.
```
or
```sh
#! /usr/bin/env nix-shell
#! nix-shell -i bash -p coreutils haskellPackages.cabal2nix-unstable git -I nixpkgs=.
#! nix-shell -i bash ci -A nixfmt
```

infinisil replied:
Alright I now removed all `nixfmt-rfc-style`'s from update script shebangs: https://github.com/NixOS/nixpkgs/pull/425068/commits/a423175f837a81be643cc660a4769d4386f53fcc

Furthermore, I updated `maintainers/scripts/update.nix` to run the `updateScript` in a development shell, which most notably should ensure that the regular automated updates don't break, because it uses that script as well: https://github.com/nix-community/nixpkgs-update/blob/f148a5d315eaf7c6985a50a05d94629022ac4812/src/Nix.hs#L290

https://github.com/NixOS/nixpkgs/pull/425068/commits/45ae92f932c542775664f86230e0d85c123a950a

MattSturgeon commented on pkgs/by-name/fl/flow-control/update.sh:
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env nix-shell
-#!nix-shell -i bash -p bash common-updater-scripts gnused nixfmt-rfc-style
+#!nix-shell -i bash -p bash common-updater-scripts gnused nixfmt
```
This shebang is using the impure `NIX_PATH` with no `-I nixpkgs=.`; so packages will come from whatever nixpkgs channel happens to be configured on the user's system.

MattSturgeon commented on pkgs/by-name/ni/nixfmt-tree/package.nix:
We also need to update `meta.longDescription` in this file.

Maybe there are other places treewide that still refer to `nixfmt-rfc-style` in comments, strings, docs, etc?

infinisil replied:
The one in `ci/default.nix` seems fine, but the others are intentional:
- `check-nix-code-blocks` is used in CI to ensure the codebase is formatted with the right version, so it's still using the older pinned CI Nixpkgs. If this is changed, the check fails
- The others are documentation, which should stay at `nixfmt-rfc-style` for now, because not everybody has a `nixfmt` that points to the rfc style version, older versions still point to the classic one.

infinisil replied:
> Ah my mistake. Does that have `allowAliases` disabled, or will we just need to _remember_ to switch when we bump the pinned nixpkgs?

I believe we'd get the delayed warning at some point, doesn't look like it disables aliases right now.

> Good point. I guess this should follow a similar timeframe to the delayed warning?
> 
> What's the best way to track doing that; should we open an issue on the formatting project board about updating documentation after a certain date?

Yeah that kind of timeline. Seems reasonable.

MattSturgeon replied:
> > Good point. I guess this should follow a similar timeframe to the delayed warning?
> > What's the best way to track doing that; should we open an issue on the formatting project board about updating documentation after a certain date?
> 
> Yeah that kind of timeline. Seems reasonable.

I've opened https://github.com/NixOS/nixpkgs/issues/425583

MattSturgeon commented on pkgs/by-name/ni/nixfmt/update.sh:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p cabal2nix curl jq
```
Looks like we're also not specifying a nixpkgs revision in our own update script.

If there's no easy way to get the pinned nixpkgs here, should we at least have `-I nixpkgs=.` like many other scripts have?

The downside of using `.` is we're more likely to depend on a package that isn't built by hydra, so perhaps using the user's NIX_PATH is preferable after all.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixfmt-1-0-0-is-out/66537/1

infinisil committed:
treewide: Remove nixfmt-rfc-style from update script shebangs

Based on the Nixpkgs used and the version of nixfmt-rfc-style in that
version, it's likely that not the correct version is used.

Update scripts should instead run within a Nixpkgs development shell
(`nix-shell`/`nix develop`/`direnv`), where the correct version of
`nixfmt` (although `treefmt` should be preferred) is always available.

infinisil committed:
maintainers/scripts/update.nix: Run updateScript in development shell

Update scripts aren't always run in the development shell by default,
especially not in the regular automated updates.
The parent commit makes the update scripts depend on running in the
development shell, so let's make sure it's always done.

MattSturgeon commented on pkgs/by-name/ri/river/update.sh:
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env nix-shell
-#!nix-shell -i bash -p bash common-updater-scripts gnused nixfmt-rfc-style zon2nix
```
Was `zon2nix` removed by mistake?

infinisil replied:
Oh my, good catch, fixed!

MattSturgeon reviewed (approved):
Thanks for such a thorough PR, and for handling the nixfmt stable release!

wolfgangwalther reviewed (approved):
Just a nit, otherwise LGTM.

infinisil commented on pkgs/README.md:
```diff
@@ -940,7 +940,9 @@ Update scripts are to be invoked by the [automatic package update script](../mai
 `update.nix` offers several modes for selecting packages to update, and it will execute update scripts for all matched packages that have an `updateScript` attribute.
 
-Each update script will be passed the following environment variables:
+Update scripts will be run inside the [Nixpkgs development shell](../shell.nix),
+providing access to some useful tools for CI.
+Furthermore each update script will be passed the following environment variables:
```
This is actually intensional. The origin of the "one sentence per line" convention is [Semantic Line Breaks](https://sembr.org/), which specifies:

> 4. A semantic line break must occur after a sentence, as punctuated by a period (.), exclamation mark (!), or question mark (?).
> 5. A semantic line break should occur after an independent clause as punctuated by a comma (,), semicolon (;), colon (:), or em dash (—).

So it's really "at most one sentence", not "exactly one sentence", which makes sense considering the motivation of making it easier to review :)

MattSturgeon replied:
I held myself back from this same nit, on the basis that there are plenty of other "offenses" in this file and the release notes usually get a formatting pass later by the release team.

Still, I'm glad to see compliance with the rule.

I agree the rule is a little heavy handed, currently. On the one hand, that may make potential automation/validation easier. On the other hand, sometimes short sentences really don't need to be on multiple lines.

philiptaron reviewed (approved):
Thanks for your years of work getting this project to 1.0, Silvan 🥇 

The piece most likely to get some future work, in my view, is the `maintainers/scripts/update.py`. Running in the CI shell is a powerful new change.

philiptaron commented:
Is there anything that now prevents a merge?

infinisil commented:
I actually don't think we want to backport this PR

Why did nixpkgs-ci add the backport label automatically?

MattSturgeon commented:
> Why did nixpkgs-ci add the backport label automatically?

The pr touches files in `ci/`.

IIUC, all changes to `ci/` and `.github/` are (almost?) always backported to _a)_ ensure things stay in sync and _b)_ make future backports easier.

infinisil commented:
Ah right. I guess this is a rare case where it's not needed :)

infinisil commented:
In a couple days I'll make the PR to update CI to use the stable version by updating the pinned Nixpkgs. That will be backported to ensure every Nixpkgs user uses the same version.

But I don't see a reason for that to go beyond Nixpkgs formatting.

MattSturgeon commented:
I think it is fine not to backport, because we've delayed the deprecation warning. No warning will occur until this current cycle becomes the oldest supported version.

If we were introducing the warning during sooner, _then_ I would expect a backport. However, doing so would raise questions about backporting breaking changes to an already released version.

emilazy commented:
I'm a bit worried about the effects of not having a formatter on stable that reflects the Nixpkgs formatting, not just for Nixpkgs but also for third-party projects with less elaborate pinning mechanisms in CI.

panchoh commented:
> I'm a bit worried about the effects of not having a formatter on stable that reflects the Nixpkgs formatting, not just for Nixpkgs but also for third-party projects with less elaborate pinning mechanisms in CI.

Agreed. My thinking is: the protocol in place (proper timely deprecation notice, etc) serves us well for most packages, but this is a special case, because it directly affects how contributors to nixpkgs and the rest of the nix ecosystem will happen. I think that, in this case, making the change in _both_ unstable and stable as quickly as possible will in fact minimize the impact, even if it causes some disruption.

I'd say, let's think this through before commiting to a standard deprecation cycle, lest we are in for a few months of chaos & madness¹, before the next stable release.

¹: I'm exaggerating, of course, but I'm really afraid of the split-brain experience we might have.

At any rate, sorry if I'm missing something obvious, just humbly trying to contribute to this awesome project. <3
